File: app/src/main/java/com/project/thisappistryingtomakeyoubetter/util/TaskViewModel.java
package com.project.thisappistryingtomakeyoubetter.util;

import android.app.Application;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.project.thisappistryingtomakeyoubetter.model.Task;

import java.util.Date;
import java.util.List;

public class TaskViewModel extends AndroidViewModel implements ViewModelProvider.Factory {

    private final Application application;
    private final Date from, to;
    private final TaskRepository taskRepository;
    private final LiveData<List<Task>> allTasks;

    public TaskViewModel(@NonNull Application application, Date from, Date to) {
        super(application);
        this.application = application;
        this.from = from;
        this.to = to;
        taskRepository = new TaskRepository(application, from, to);
        allTasks = taskRepository.getTasks();
    }

    public LiveData<List<Task>> getTasks() {
        return allTasks;
    }

    public void insert(Task task) {
        taskRepository.insert(task);
    }

    public void update(Task task) {
        taskRepository.update(task);
    }

    public void delete(Task task) {
        taskRepository.delete(task);
    }

    @SuppressWarnings("unchecked")
    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        return (T) new TaskViewModel(application, from, to);
    }
}

File: app/src/main/java/com/project/thisappistryingtomakeyoubetter/util/TaskViewModel.kt
package com.project.thisappistryingtomakeyoubetter.util

import android.util.Log
import androidx.lifecycle.*
import com.project.thisappistryingtomakeyoubetter.model.*
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.launch
import java.util.*
import javax.inject.Inject

class TaskViewModel @Inject constructor(
    private val taskRepository: TaskRepository,
    private val labelRepository: LabelRepository
) : ViewModel() {

    private val from = MutableLiveData<Date?>()
    fun setFrom(from: Date?) {
        this.from.value = from
    }

    private val to = MutableLiveData<Date?>()
    fun setTo(to: Date?) {
        this.to.value = to
    }


    val tasksWithLabel: LiveData<List<TaskWithLabel>?> = Transformations.switchMap(from) { from ->
        Transformations.switchMap(to) { to ->
                taskRepository.getTaskWithLabel(from, to)
        }
    }

    val taskGroup: LiveData<List<TaskGroup>> = Transformations.switchMap(tasksWithLabel){ list ->
        val result = MutableLiveData<List<TaskGroup>>()
        val taskGroup = mutableListOf<TaskGroup>()
        val map = mutableMapOf<Date, List<TaskWithLabel>>()

        list?.forEach {
            // using map, find the associate date then assign value with list
            val date = it.task.date?:Date()
            val a = map.get(date)
            if(a == null) {
                map[date] = listOf(it)
            } else {
                val b = a.toMutableList()
                b.add(it)
                map[date] = b
            }
        }

        map.forEach {
            taskGroup.add(TaskGroup(it.key, it.value))
        }
        result.value = taskGroup
        result
    }

    val label: LiveData<List<Label>?> = labelRepository.getAll()


    fun insert(task: Task) {
        viewModelScope.launch(IO) { taskRepository.insert(task) }
    }

    fun update(task: Task) {
        viewModelScope.launch(IO) { taskRepository.update(task) }
    }

    fun delete(task: Task) {
        viewModelScope.launch(IO) { taskRepository.delete(task) }
    }

    fun deleteAll() {
        viewModelScope.launch(IO) { taskRepository.deleteAll() }
    }

}